### PR TITLE
Added feature to siphon off data as list

### PR DESCRIPTION
**Pull request by sgnoohc**
- changed the histogram outputting to ```dict_accumulator```
- Also added ```list_accumulator``` if option ```--siphon``` is used

### CODE REVIEW

File: analysis/wwz/wwz4l.py [hunk 1]
```diff
@@ -1045,9 +1076,28 @@ def process(self, events):
...
                             "systematic"    : wgt_fluct,
                         }
-                        hout[dense_axis_name].fill(**axes_fill_info_dict)
-
-        return hout
+                        self.accumulator[dense_axis_name].fill(**axes_fill_info_dict)
+
+            if self._siphon_bdt_data:
+                # Fill the list accumulator
+                self.accumulator["bdt_of_wwz_list"] += bdt_of_wwz[sr_4l_bdt_of_trn].to_list()
```
Comment (kmohrman) [inline analysis/wwz/wwz4l.py:1083]: Would it make sense to set this part up as a for loop so we don't have to do all of these lines of `+=`?

File: analysis/wwz/wwz4l.py [hunk 2]
```diff
@@ -140,14 +140,54 @@ def __init__(self, samples, wc_names_lst=[], hist_lst=None, ecut_threshold=None,
...
+                "bdt_sf_bkg_list",
+                "bdt_sf_proc_list",
+                "bdt_sf_wgt_list",
+                "bdt_sf_evt_list",
+            ]
+            for list_output_name in list_output_names:
+                dout[list_output_name] = processor.list_accumulator([])
+
+            # Also adding BDT input variable list accumulators
+            for of_bdt_var_name in get_ec_param("of_bdt_var_lst"):
```
Comment (kmohrman) [inline analysis/wwz/wwz4l.py:175]: The input lists are nearly identical. Would it make sense to just make a `set` of the two lists? Or is there a reason why you want to specifically have the OF and SF ones kept track of separately?

Comment (sgnoohc) [reply analysis/wwz/wwz4l.py:175]: Because the set of events that fall into each branch are different

File: analysis/wwz/wwz4l.py (hunk 1 above)
Comment (sgnoohc) [reply analysis/wwz/wwz4l.py:1083]: how do you do that when the behaviors are different? if you do a for loop it's gonna have ugly if statements

File: analysis/wwz/wwz4l.py (hunk 2 above)
Comment (kmohrman) [reply analysis/wwz/wwz4l.py:175]: Ahh, ok, got it. I was not looking at how they get used.

File: analysis/wwz/wwz4l.py (hunk 1 above)
Comment (kmohrman) [reply analysis/wwz/wwz4l.py:1083]: Hmm, it looks like most of the lines are fairly similar. OF and SF seem to do the same thing. I will try to think about it a bit more.

**Comment by kmohrman** [comment]
I've checked the yields and cards and significance and indeed things seem to match (so the change to the dict accumulator seems to be validate). I guess you've probably already checked that too. 

One question though, I'm wondering about the plan for including code for interacting with the "siphoned" BDT info. Would you see that as part of ewkcoffea, or a standalone repository? Is it currently committed anywhere?

**Comment by sgnoohc** [comment]
It currently just lives in my local area uncommitted. I have two scripts, "ingress.py" that prepares data in pkl, and "train.py" that trains bdt using XGBoost and writes out xgboost model. Would you like me to commit?

**Comment by kmohrman** [comment]
> It currently just lives in my local area uncommitted. I have two scripts, "ingress.py" that prepares data in pkl, and "train.py" that trains bdt using XGBoost and writes out xgboost model. Would you like me to commit? I have no plans for it. If matthew does bdt for Run 3 I can hand him the code.

I think it would make sense to commit it. Could put it in the analysis/wwz directory, or could make a new sub dir under analysis for it (something like analysis/bdt_training or something).

**Comment by sgnoohc** [comment]
I just added the data ingressing and training script

**Comment by sgnoohc** [comment]
The Lint error on "mpltern". I do not know what to do. Because mpltern MUST be imported for "ternary" to be recognized within matplotlib. Without that line it will throw an error. but the script does not directly call "mpltern. something"

**Comment by kmohrman** [comment]
Are there additional packages you had to conda (or pip) install to run the scrips in the bdt training directory? Could you provide the list and/or push the extra ones to the environment yaml file? I was trying to run your training script (just to test the imports you mentioned that flake8 was complaining about) but got a couple errors about missing packages.

**Comment by sgnoohc** [comment]
What was the exact error?

--
Dr. P. Chang, Asst. Prof., University of Florida
***@***.***
________________________________

**Comment by kmohrman** [comment]
A couple `ModuleNotFoundError` on `mpltern` and `sklearn`

**Comment by sgnoohc** [comment]
conda install -c conda-forge mpltern
conda install -c conda-forge scikit-learn

--
Dr. P. Chang, Asst. Prof., University of Florida
***@***.***
________________________________

**Comment by kmohrman** [comment]
Right, but would want to add these to the yaml then, I think? Are there any others you had to also install?

**Comment by kmohrman** [comment]
Note, in addition to a bunch of minor updates to the yield scripts, there are a few more substantial changes that were applied on this branch:
* Bug fix for R3 sample grouping (WZ had been listed by itself and in "other")
* Commenting out tWll samples in 4l skim json. Does not impact our results because we were specifically grabbing the tWZ 4l samples, but it is unnecessary to process them (and could lead to downstream confusion)
* Updating R3 ref in the ref yields py file (which had not been updated since before the ID changes and such, so the changes are not tiny)
* Modifying the processor to do the BDT calculation for R3 same as R2. Eventually we'll swap in the R3 BDT when we have it.

**Comment by kmohrman** [comment]
Also to note for future reference: There were significant performance issues observed when running this branch with WQ. However, at the same time as these tests, HPG was starting to have some major unrelated issues. At the present time, it is essentially impossible to run on HPG, so the possible WQ performance issues are unconfirmed (though it is possible to speculate that if they are true, they are related to the change in the way the accumulation is being done in the processor). It is also possible (though not checked) that the environment is being rebuilt unnecessarily on this branch for some reason (i.e. something may be broken in the way the remote environment tool from topcoffea checks to see if the env needs to be rebuilt). Though since we are running entirely with futures right now (at UCSD) this has not been fully explored or confirmed.

**Pull request merged**